fix(router): do not finish bootstrap until all the routes are resolved

vsavkin opened the pull request:
**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

The bootstrap doesn't wait for the router to finish the initial navigation.

**What is the new behavior?**

There is an option you can pass to RouterModule.foRoot to make bootstrap wait. Since many depend on the current behavior, it is preserved as a default and is deprecated.


**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**:

vicb commented on modules/@angular/router/src/router_module.ts:
```diff
@@ -208,6 +210,26 @@ export function provideRoutes(routes: Routes): any {
+ * * 'legacy_enabled'- the initial navigation starts after the root component has been created.
+ * The bootstrap is not blocked until the initial navigation is complete. @deprecated
+ * * 'legacy_disabled'- the initial navigation is not performed. The localtion listener is set up
+ * after @deprecated
```
- typos in "localtion"
- could you be more verbose on what the preferred choice should be (ie most probably `"enabled"`) and when you would like to switch to "disable". Also say that the "legacy_..." options are for backward compatibility and should not be used for new apps ?

vicb commented on modules/@angular/router/src/router_module.ts:
```diff
@@ -208,6 +210,26 @@ export function provideRoutes(routes: Routes): any {
+ * * 'legacy_disabled'- the initial navigation is not performed. The localtion listener is set up
+ * after @deprecated
+ * the root component gets created.
+ * * true - same as 'legacy_enabled'. @deprecated
```
back quote true/false

vicb commented on modules/@angular/router/src/router_module.ts:
```diff
@@ -278,22 +300,100 @@ export function rootRoute(router: Router): ActivatedRoute {
+    this.resultOfPreactivationDone.complete();
+  }
+
+  isLegacyEnabled(opts: ExtraOptions): boolean {
```
move these 2 mehtods out of the class ?

vicb commented on modules/@angular/router/src/router_module.ts:
```diff
@@ -278,22 +300,100 @@ export function rootRoute(router: Router): ActivatedRoute {
+ */
+@Injectable()
+export class RouterInitializer {
+  private initNavigation: boolean;
```
`initNavigationDone = false` ?

vicb commented:
Please add a `DEPRECATIONS` section in the commit message. Thanks !

vsavkin committed:
fix(router): do not finish bootstrap until all the routes are resolved

DEPRECATION

Use `RouterModule.forRoot(routes, {initialNavigation: 'enabled'})` instead of
`RouterModule.forRoot(routes, {initialNavigtaion: true})`.

Before doing this, move the initialization logic affecting the router
from the bootstrapped component to the boostrapped module.

Similarly, use `RouterModule.forRoot(routes, {initialNavigation: 'disabled'})`
instead of `RouterModule.forRoot(routes, {initialNavigation: false})`.

Deprecated options: 'legacy_enabled', `true` (same as 'legacy_enabled'),
'legacy_disabled', `false` (same as 'legacy_disabled').

The "Router Initial Navigation" design document covers this change.
Read more here:
https://docs.google.com/document/d/1Hlw1fPaVs-PCj5KPeJRKhrQGAvFOxdvTlwAcnZosu5A/edit?usp=sharing

IgorMinar commented on tools/public_api_guard/router/typings/router.d.ts:
```diff
@@ -76,7 +76,7 @@ export declare type Event = NavigationStart | NavigationEnd | NavigationCancel |
     enableTracing?: boolean;
     errorHandler?: ErrorHandler;
-    initialNavigation?: boolean;
+    initialNavigation?: InitialNavigation;
```
isn't this a breaking change? when was this interface introduced?

vicb replied:
see https://github.com/angular/angular/pull/14762/files#diff-c0baae5e1df628e1a217e8dc38557fcbR237

the changes are described in the commit comment: former options are valid and have the same meaning (`true`/`false`), new values have a new meaning and are documented,
